README.md
# Лабораторная работа №1 по дисциплине "Компьютерная графика"

[![Version](https://img.shields.io/badge/Version-1.0.0-blue.svg)](https://github.com/Simbirkinaa/cg_lab1)
![Python](https://img.shields.io/badge/Python-3.9-pink)
[![Build, Test](https://github.com/Krnvaa/Drawing_nested_rectangles/actions/workflows/python-app.yml/badge.svg)](https://github.com/Krnvaa/Drawing_nested_rectangles/actions/workflows/python-app.yml)
[![License: MIT ](https://img.shields.io/badge/License-MIT-fuchsia.svg)](https://opensource.org/licenses/MIT)
[![Telegram](https://img.shields.io/badge/Telegram-@milayavafelka-blue?style=social&logo=telegram)](https://t.me/milayavafelka)

![pushincat](https://github.com/Simbirkinaa/cg_lab1/blob/master/gif%26img/photo_2023-09-30_10-31-25.jpg)
## Описание задачи
Требуется составить программу, которая выводит на экран треугольник с заданными координатами вершин, заполненный заданным образцом. Затем при каждом нажатии клавиш управления 
курсором осуществляла бы перемещение треугольника на экране с заданным шагом в соответствующем направлении. При каждом нажатии клавиши Enter
изменяла бы образец заполнения треугольника на следующий циклически. Выход 
из программы по нажатию клавиши ESC.
## Демонстрация выполнения работы программы
![GIF](https://github.com/Simbirkinaa/cg_lab1/blob/master/gif%26img/треугольник-_1_.gif)

## Программа реализована на языке:

| Язык | Интерпретатор/Версия | Среда разработки | 
| ------ | ------ | ------ |
| Python | Python /   3.9  | PyCharm 2022.2.3 |

## Установка и запуск
1. ```git clone https://github.com/Simbirkinaa/cg_lab1.git```
2. ```cd <project_folder>```
3. ```python main.py```


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [1.0.0] - 2023-09-23

### Added
- A graphical interface has been created with the ability to control a triangle.
- The drawing of a triangle with various samples has been implemented.
- Added the ability to move the triangle using the control keys.
- Switching between triangle samples is implemented by pressing the Enter key.
- Added the ability to terminate the program by pressing the Escape key.

### Changed
- Added support for inverted triangle patterns.
- Improvements have been made to the graphical interface, including changing the color scheme

### Fixed

- Fixed a bug with updating the triangle sample when pressing the Enter key.
- Fixed a bug when the triangle went beyond the canvas borders.


Tests/unittest1.py
import unittest
import tkinter as tk
from tkinter import Canvas
from main import TriangleDrawer

class TestTriangleDrawer(unittest.TestCase):

    def setUp(self):
        self.root = tk.Tk()
        self.canvas = Canvas(self.root, width=300, height=300)
        self.canvas.pack()
        self.triangle_drawer = TriangleDrawer(self.canvas)

    def test_initial_triangle_pattern(self):
        # Проверка, что начальный образец треугольника отображается правильно

        objects_on_canvas = self.canvas.find_all()
        self.assertTrue(len(objects_on_canvas) > 0, "На холсте нет объектов")

    def test_move_triangle(self):
        # Проверка перемещения треугольника вправо

        initial_x = self.triangle_drawer.x
        event = tk.Event()
        event.keysym = "Right"
        self.triangle_drawer.move_triangle(event)
        expected_x = initial_x + self.triangle_drawer.step
        self.assertEqual(self.triangle_drawer.x, expected_x, "Координата x изменилась неправильно")

    def test_change_pattern(self):
        # Проверка изменения образца
        initial_pattern_index = self.triangle_drawer.pattern_index
        event = tk.Event()
        self.triangle_drawer.change_pattern(event)
        expected_pattern_index = (initial_pattern_index + 1) % len(self.triangle_drawer.triangle_patterns)
        self.assertEqual(self.triangle_drawer.pattern_index, expected_pattern_index,
                         "Индекс образца изменился неправильно")

    def tearDown(self):
        self.root.destroy()



if __name__ == '__main__':
    unittest.main()


main.py
import tkinter as tk
class TriangleDrawer:
    def __init__(self, canvas):
        # Этот конструктор инициализирует объект класса и принимает один аргумент - холст canvas.
        # Холст представляет собой область для рисования. Внутри конструктора устанавливаются
        # начальные значения для различных параметров:
        # self.canvas: Ссылка на холст, на котором будет отображаться треугольник.
        # self.x и self.y: Начальные координаты треугольника на холсте.
        # self.step: Шаг перемещения при нажатии клавиш управления.
        # self.triangle_patterns: Список образцов треугольников, представленных в виде координат линий.
        # self.pattern_index: Текущий индекс выбранного образца треугольника (начинается с первого образца).

        self.canvas = canvas
        self.x, self.y = 120, 150
        self.step = 10
        self.triangle_patterns = [
    [
        (0, 0, 50, 0),
        (0, 0, 25, -50),
        (50, 0, 25, -50)
    ],
    [
        (0, 0, 50, 0),
        (0, 0, -25, 50),
        (50, 0, -25, 50)
    ],
    [
        (0, 0, 50, 0),
        (0, 0, 25, 50),
        (50, 0, 25, 50)
    ],
    [
        (0, 0, 50, 0),
        (50, 0, 75, 50),
        (0, 0, 75, 50)
    ],
]
        self.pattern_index = 0
        self.draw_triangle()

    def draw_triangle(self):
        # Этот метод отвечает за отображение треугольника на холсте.
        # Сначала он удаляет все предыдущие элементы с тегом "triangle" с холста.
        # Затем для текущего выбранного образца треугольника, он создает линии на холсте, используя координаты
        # из образца. Эти линии представляют собой отрезки, которые образуют треугольник.
        # Цвет линий устанавливается как "black", а ширина линий - 2 пикселя.

        self.canvas.delete("triangle")
        for coords in self.triangle_patterns[self.pattern_index]:
            self.canvas.create_line(
                self.x + coords[0], self.y + coords[1], self.x + coords[2], self.y + coords[3],
                fill="black", width=2, tags="triangle"
            )

    def move_triangle(self, event):
        #Этот метод обрабатывает события перемещения треугольника в ответ на нажатия клавиш управления курсором
        # (вверх, вниз, влево, вправо).	В зависимости от нажатой клавиши, метод изменяет координаты треугольника
        # на холсте, добавляя или вычитая значение шага (self.step) к self.x или self.y.
        # После изменения координат вызывается метод draw_triangle() для обновления отображения треугольника.

        step = 10  # Шаг перемещения
        if event.keysym == "Right":
            self.x += step
        elif event.keysym == "Left":
            self.x -= step
        elif event.keysym == "Down":
            self.y += step
        elif event.keysym == "Up":
            self.y -= step
        self.draw_triangle()

    def change_pattern(self, event):
        # Этот метод обрабатывает событие смены образца треугольника при нажатии клавиши Enter.
        # Он увеличивает значение self.pattern_index на 1, переходя к следующему образцу в списке.
        # Если конец списка образцов достигнут, индекс обнуляется и выбирается первый образец.
        # После изменения образца вызывается метод draw_triangle() для его отображения.

        self.pattern_index = (self.pattern_index + 1) % len(self.triangle_patterns)
        self.draw_triangle()

    def exit_program(self, event):
        # Метод exit_program обрабатывает событие завершения работы приложения при нажатии клавиши Escape.
        # Он вызывает метод root.quit(), который завершает выполнение программы.
        if event.keysym == "Escape":
            root.quit()

root = tk.Tk()
root.title("Управление треугольником")
canvas = tk.Canvas(root, width=300, height=300)
canvas.pack()
triangle_drawer = TriangleDrawer(canvas)

# <Right>, <Left>, <Down>, <Up>:
# Эти строки устанавливают обработчики для клавиш управления курсором (вправо, влево, вниз, вверх).
# Когда пользователь нажимает одну из этих клавиш, вызывается метод move_triangle,
# который отвечает за перемещение треугольника в соответствующем направлении.


# <Return> устанавливает обработчик для клавиши Enter.
# При нажатии клавиши Enter вызывается метод change_pattern, который отвечает за смену образца треугольника.
# <Escape> устанавливает обработчик для клавиши Escape.
# Когда пользователь нажимает клавишу Escape, вызывается метод exit_program, который завершает работу программы.
root.bind("<Right>", triangle_drawer.move_triangle)
root.bind("<Left>", triangle_drawer.move_triangle)
root.bind("<Down>", triangle_drawer.move_triangle)
root.bind("<Up>", triangle_drawer.move_triangle)
root.bind("<Return>", triangle_drawer.change_pattern)
root.bind("<Escape>", triangle_drawer.exit_program)

root.mainloop()